Userland/Libraries/LibSemVer/SemVer.h

```diff
@@ -0,0 +1,107 @@
+#include <AK/Types.h>
+
+namespace SemVer {
+enum class BumpType {
+    Major,
+    Minor,
+    Patch,
+    Prerelease,
+};
+
+enum class CompareType {
+    Exact,
+    Major,
+    Minor,
+    Patch
+};
+
+class SemVer {
+    char const m_number_separator;
+    u64 m_major;
+    u64 m_minor;
+    u64 m_patch;
+    Vector<String> m_prerelease_identifiers;
+    String m_buildmetadata;
+
+public:
+    SemVer(u64 major, u64 minor, u64 patch, char m_number_separator)
+        : m_number_separator(m_number_separator)
+        , m_major(major)
+        , m_minor(minor)
+        , m_patch(patch)
+    {
+    }
+
+    SemVer(u64 major, u64 minor, u64 patch, char m_number_separator, Vector<String> const& prereleases, String const& buildmetadata)
+        : m_number_separator(m_number_separator)
+        , m_major(major)
+        , m_minor(minor)
+        , m_patch(patch)
+        , m_prerelease_identifiers(prereleases)
+        , m_buildmetadata(buildmetadata)
+    {
+    }
+
+    [[nodiscard]] u64 major() const { return m_major; }
+    [[nodiscard]] u64 minor() const { return m_minor; }
+    [[nodiscard]] u64 patch() const { return m_patch; }
+    [[nodiscard]] ReadonlySpan<String> prerelease_parts() const { return m_prerelease_identifiers.span(); }
+    [[nodiscard]] String prerelease() const
+    {
+        StringBuilder builder;
+        builder.join('.', m_prerelease_identifiers);
+        return MUST(builder.to_string());
+    }
+    [[nodiscard]] String buildmetadata() const { return m_buildmetadata; }
+
+    [[nodiscard]] ErrorOr<SemVer> bump(BumpType type) const;
+
+    [[nodiscard]] bool is_same(SemVer const& other, CompareType compare_type = CompareType::Exact) const;
+    [[nodiscard]] bool is_greater_than(SemVer const& other) const;
+    [[nodiscard]] bool is_lesser_than(SemVer const& other) const { return !is_same(other) && !is_greater_than(other); }
+    [[nodiscard]] bool operator==(SemVer const& other) const { return is_same(other); }
+    [[nodiscard]] bool operator!=(SemVer const& other) const { return !is_same(other); }
+    [[nodiscard]] bool operator>(SemVer const& other) const { return is_lesser_than(other); }
+    [[nodiscard]] bool operator<(SemVer const& other) const { return is_greater_than(other); }
+    [[nodiscard]] bool operator>=(SemVer const& other) const { return *this == other || *this > other; }
+    [[nodiscard]] bool operator<=(SemVer const& other) const { return *this == other || *this < other; }
+
+    [[nodiscard]] bool satisfies(StringView const& spec) const;
+    [[nodiscard]] String suffix() const;
+    [[nodiscard]] String to_string() const;
+};
+
+/**
+ * Valid semver formats as per v2.0.0 spec
+ * <valid semver> ::= <version core>
+ *                 | <version core> "-" <pre-release>
+ *                 | <version core> "+" <build>
+ *                 | <version core> "-" <pre-release> "+" <build>
+ *
+ * Because of size_t overflow, it won't work work version such as 99999999999999999999999.
+ */
```

Comment:
Wrong comment format.

Reply:
If I wanted to add the comments here, should I use multiple `//` comments or it is not allowed here?

Reply:
Nvm, moved the comment to Test case.